library/std/src/sys/pal/unix/fs.rs

```diff
@@ -739,44 +739,37 @@ impl Iterator for ReadDir {
                 // Like for uninitialized contents, converting entry_ptr to `&dirent64`
-                // would not be legal. However, unique to dirent64 is that we don't even
-                // get to use `&raw const (*entry_ptr).d_name` because that operation
-                // requires the full extent of *entry_ptr to be in bounds of the same
-                // allocation, which is not necessarily the case here.
-                //
-                // Instead we must access fields individually through their offsets.
-                macro_rules! offset_ptr {
-                    ($entry_ptr:expr, $field:ident) => {{
-                        const OFFSET: isize = mem::offset_of!(dirent64, $field) as isize;
-                        if true {
-                            // Cast to the same type determined by the else branch.
-                            $entry_ptr.byte_offset(OFFSET).cast::<_>()
-                        } else {
-                            #[allow(deref_nullptr)]
-                            {
-                                &raw const (*ptr::null::<dirent64>()).$field
-                            }
-                        }
+                // would not be legal. However, we can use `&raw const (*entry_ptr).d_name`
+                // to refer the fields individually, because that operation is equivalent
+                // to `byte_offset` and thus does not require the full extent of `*entry_ptr`
+                // to be in bounds of the same allocation, only the offset of the field
+                // being referenced.
+                macro_rules! entry_field_ptr {
+                    ($field:ident) => {{
+                        // To make sure the field actually exists and is visible,
+                        // and we aren't silently doing any Deref coercion.
+                        const _: usize = mem::offset_of!(dirent64, $field);
+                        &raw const (*entry_ptr).$field
                     }};
                 }
 
                 // d_name is guaranteed to be null-terminated.
-                let name = CStr::from_ptr(offset_ptr!(entry_ptr, d_name).cast());
+                let name = CStr::from_ptr(entry_field_ptr!(d_name).cast());
                 let name_bytes = name.to_bytes();
                 if name_bytes == b"." || name_bytes == b".." {
                     continue;
                 }
 
                 #[cfg(not(target_os = "vita"))]
                 let entry = dirent64_min {
-                    d_ino: *offset_ptr!(entry_ptr, d_ino) as u64,
+                    d_ino: *entry_field_ptr!(d_ino) as u64,
                     #[cfg(not(any(
                         target_os = "solaris",
                         target_os = "illumos",
                         target_os = "aix",
                         target_os = "nto",
                     )))]
-                    d_type: *offset_ptr!(entry_ptr, d_type) as u8,
+                    d_type: *entry_field_ptr!(d_type) as u8,
```

Comment:
Any idea why the `as` casts are here? Looking at `libc` the ino types are just typedefs so it seems they wouldn't be needed. https://github.com/rust-lang/libc/blob/1867bf30eb9d22d3f8e86c2a8c25cf92fe5580da/src/unix/linux_like/linux/mod.rs#L1474-L1480 https://github.com/rust-lang/libc/blob/1867bf30eb9d22d3f8e86c2a8c25cf92fe5580da/src/unix/linux_like/linux/mod.rs#L11 https://github.com/rust-lang/libc/blob/1867bf30eb9d22d3f8e86c2a8c25cf92fe5580da/src/unix/mod.rs#L9

Reply:
I don't know why the `as` casts are there, no. `dirent64` is imported as aliases from various different platforms, so maybe the `as` casts were added just to make sure it worked (or at least compiled) on any weird platforms where the fields were different type?

Removing the `as` casts doesn't make any platforms I've tried with `./x.py check library/std` complain though[^1], so maybe they could be removed?

[^1]: `x86_64-unknown-linux-gnu`, `aarch64-unknown-linux-gnu`, `aarch64-unknown-freebsd`, `arm-linux-androideabi`, `armv7-linux-androideabi`, `i686-linux-android`